### ISSUE TITLE
Provide ability to specify different cache profile based on status code

### ISSUE THREAD

**Issue by alefranz** [aspnet/AspNetCore#9920]
### Is your feature request related to a problem? Please describe.
I'm trying to prevent user agents to cache errors or 404s.

### Describe the solution you'd like
I think it should be possible to provide a different cache behaviour (either with cache profiles or `ResponseCacheAttribute`) based on the status code.

### Describe alternatives you've considered
The current solution is to not use the provided `ResponseCacheFilter` but re-implement one like
```csharp
    public class CachingFilterAttribute : Attribute, IResultFilter
    {
        public void OnResultExecuting(ResultExecutingContext context)
        {
            if (context.Result is StatusCodeResult statusCodeResult &&
                statusCodeResult.StatusCode == 200)
            {
                context.HttpContext.Response.GetTypedHeaders().CacheControl =
                    new CacheControlHeaderValue
                    {
                        Public = true,
                        MaxAge = TimeSpan.FromHours(1)
                    };
                context.HttpContext.Response.Headers[HeaderNames.Vary] = new[] {"Accept-Encoding,Accept"};
            }
        }

        public void OnResultExecuted(ResultExecutedContext context)
        {
        }
    }
```

### Additional context
Also related to https://github.com/aspnet/AspNetCore.Docs/issues/11990 as the use of the ResponseCacheMiddleware requires to add the cache headers. The middleware then doesn't cache non 200 responses, however it is not immediately obvious that the header will be left even for error responses and the user agents will cache it.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @alefranz.
@javiercn  can you please look into this? Thanks!

**Comment by alefranz** [comment]
Hello!

I had a look at how this could be implemented and I believe this would require some fundamental changes:
- The logic in the `ResponseCacheFilter` is executed in `OnActionExecuting` so the status code is not yet available.
- Adding a conditional in the Response Cache (e.g. only if 200) could probably be counter intuitive without the ability to cascade to the next rule that match, but that would require fundamental changes:
  - currently the `ResponseCacheAttribute` can be applied only once 
https://github.com/aspnet/AspNetCore/blob/5502c2080e9c48e05effdb91695506cdbc14601a/src/Mvc/Mvc.Core/src/ResponseCacheAttribute.cs#L16
  - Only one policy is applied https://github.com/aspnet/AspNetCore/blob/5502c2080e9c48e05effdb91695506cdbc14601a/src/Mvc/Mvc.Core/src/Filters/ResponseCacheFilter.cs#L90-L97

What do you think?
Would you consider a PR or is this out of scope for the ResponseCacheFilter?

Thank you,
 Alessio

**Comment by alefranz** [comment]
Hello,

I toyed with this idea and I'm working on the ability to:
- Add the ability on the `ResponseCacheAttribute` to specify a list of status codes it applies to:
  ```csharp
        /// <summary>
        /// Gets or sets the status codes for which apply the cache headers.
        /// </summary>
        /// <remarks>
        /// When not specified or empty, the headers are always added.
        /// </remarks>
        public int[] ApplyForStatusCodes { get; set; }
  ```
- Move the ResponseCacheFilter logic to `OnActionExecuted`
  This will be a breaking change, the most important difference I believe will be that it will override headers eventually set in the controller, however the documentation state 
> Removes any existing headers for Vary, Cache-Control, and Pragma.
so the new behavior will be more in line with the documentation than the current one.
- Apply the caching defined by the most specific filter which has the status code that match.

What do you think about it?
Is it something that could be considered and I should submit the PR?

Thank you,
 Alessio

**Comment by javiercn** [comment]
@alefranz Thanks for reaching out.

I'm not a huge expert on this area. Your idea sounds doable, but I'm not sure is something we would want out of  the box, as I feel there are a couple of problematic points.

> * Move the ResponseCacheFilter logic to `OnActionExecuted`
>   This will be a breaking change, the most important difference I believe will be that it will override headers eventually set in the controller, however the documentation state

I think this would be a problem, as it might lead to security issues by overriding some of the cache related headers with less secure values. For example, when we apply antiforgery, this would override it, which is something we don't want.

> * Apply the caching defined by the most specific filter which has the status code that match.

This would also be something somewhat sophisticated that would likely be hard to explain/debug in many situations.

I'm more inclined to keep the existing behavior than to introduce a change like this at this point of the release.

I do understand where the ask is coming from, but this is also something that can be implemented in a few lines of code in your app (as you demonstrated above). I'm not 100 sure what you loose by not using our default attribute.

I would also check if you are using this in combination with the response-caching middleware, as that already avoids caching non 200 responses.

I think if we were to do anything would be to restrict the set of situations where the caching headers apply to be consistent with what the response caching middle-ware does.

Hope this helps.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. We believe that the question you've raised have been answered. If you still feel a need to continue the discussion, feel free to reopen it and add your comments.
